core/src/main/java/com/google/instrumentation/tags/TagSet.java

```diff
@@ -0,0 +1,177 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.instrumentation.tags;
+
+import com.google.instrumentation.internal.StringUtil;
+import javax.annotation.Nullable;
+
+/** A set of tags. */
+public abstract class TagSet {
+  /** The maximum length for a string tag value. */
+  public static final int MAX_STRING_LENGTH = StringUtil.MAX_LENGTH;
+
+  /**
+   * Determines whether a key is present.
+   *
+   * @param key the key to look up.
+   * @return {@code true} if the key is present.
+   */
+  public abstract boolean tagKeyExists(TagKey<?> key);
+
+  // TODO(sebright): Which is better, the generic "get" method or three specialized "get" methods?
+  // Should these methods take defaults or throw exceptions to handle missing keys?
+
+  /**
+   * Gets a value of type {@code String}.
+   *
+   * @param key the key to look up.
+   * @return the tag value, or {@code null} if the key isn't present.
+   */
+  @Nullable
+  public abstract String getStringTagValue(TagKey<String> key);
+
+  /**
+   * Gets a value of type {@code long}.
+   *
+   * @param key the key to look up.
+   * @param defaultValue the value to return if the key is not preset.
+   * @return the tag value, or the default value if the key is not present.
+   */
+  public abstract long getIntTagValue(TagKey<Long> key, long defaultValue);
+
+  /**
+   * Gets a value of type {@code boolean}.
+   *
+   * @param key the key to look up.
+   * @param defaultValue the value to return if the key is not preset.
+   * @return the tag value, or the default value if the key is not present.
+   */
+  public abstract boolean getBooleanTagValue(TagKey<Boolean> key, boolean defaultValue);
+
+  /**
+   * Gets a value of any type.
+   *
+   * @param key the key to look up.
+   * @return the tag value, or {@code null} if the key isn't present.
+   */
+  @Nullable
+  public abstract <T> T getTagValue(TagKey<T> key);
+
+  /**
+   * Returns a builder based on this {@code TagSet}.
+   *
+   * @return a builder based on this {@code TagSet}.
+   */
+  public abstract Builder toBuilder();
+
+  /** Builder for the {@link TagSet} class. */
+  public abstract static class Builder {
+
+    /**
+     * Adds the key/value pair if the key is not present. If the key is present, it logs an error.
+     *
+     * @param key the key to look up.
+     * @param value the value to insert for the given key.
+     * @return this
+     */
+    public abstract Builder insert(TagKey<String> key, String value);
+
+    /**
+     * Adds the key/value pair if the key is not present. If the key is present, it logs an error.
+     *
+     * @param key the key to look up.
+     * @param value the value to insert for the given key.
+     * @return this
+     */
+    public abstract Builder insert(TagKey<Long> key, long value);
```

Comment:
We should follow java map or set (probably map) standards and use put or add instead.